Add_Keyboard_Support/DriverImplementation.md

```diff
@@ -0,0 +1,217 @@
+
+The state update is done by just updating the current_state variable. As already said that variable will start with the Normal State, so we will need an init function to e called before enabling the IRQ: 
+
+```c
+#define NORMAL_STATE 0
+#define PREFIX_STATE 1
+
+uint8_t current_state;
+
+void init_keyboard() {
+    // Do other initialization stuff like: clean the keyboard buffer, identify the scancode set, enable the IRQ etc.
+    current_state = NORMAL_STATE;
+}
+
+void keyboard_driver_irq_handler() {
+    int scancode = inb(0x60); // Read byte from the Keyboard data port
+    if (scancode == 0xE0) {
+        current_state = PREFIX_STATE
+        // We have read a prefix, so let's update the state and finish here
+        // this is a very simple scenario, there could be more needed depending on the design
+        return;
+    }
+    if (current_state == PREFIX_STATE) {
+        // Do what you need to store the key_code and eventually translate it to the kernel_code and return to the normal state
+        current_state = NORMAL_STATE;
+    }
+}
+```
+
+The code snippet above just show how the states are updated, it doesn't care about the rest. Of course the state machine we have implemente is a very simple one, but if we are going for example to add support for the Print Screen and Pause key, it will probably have more states. The buffer update code can be reused from the previous paragraph. 
+
+### Handling the "Special" keys
+
+When talking about special keys we are referring to those keys that are considered like modifiers keys, when pressed they can alter the output from other keys (we are not talking about the scancode, but for the translated key), or they are simply used in combination with other keys, we are mostly referring to the _Ctrl_, _Alt_, _Shift_, and to some extent also the caps lock (even if this little bit easier to handle). 
+
+When these keys are being kept presed the driver need to know this informantion, because it has to behave accordingly: 
+
+* If the `shift` key is pressed the translation to ascii mechanism needs to know it because it has to return the Capital Symbol associated with that key (for Capital Symbol we mean or the Letter or the symbol above some keys) 
+* If `ctrl` and `alt` are pressed the driver needs to know it because it can trigger either a key combination or some of the "Alt"ernative symbols on some special keyboard keys.
+* If the caps lock key is pressed (not kept pressed) we need the translation function to return only the Capital version of the keys associated with alphabetical letters.
+* The `fn` button is similar to `ctrl` and `alt` but it usually will just trigger an alternate "behaviour" for the keys (** TODO: check if changes the scancode generated by the key**)
```

Comment:
fn key is handled by firmware (i.e. the bios). It's not normally visible to the OS. If you do fn+ f12 for example, the os might see it as "volume up" or some other key, not "fn + f12".